File: 2-add-two-numbers/add-two-numbers.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: Optional[ListNode]
        :type l2: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        result=ListNode()
        r=result
        residue=0
        while l1 or l2 or residue:
            sum=residue
            if l1:
                sum+=l1.val
                l1=l1.next
            if l2:
                sum+=l2.val
                l2=l2.next
            residue=sum//10
            r.next=ListNode(sum%10)
            r=r.next
        return result.next


        

File: 2-add-two-numbers/add-two-numbers.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2) {
    struct ListNode* result=(struct ListNode*)malloc(sizeof(struct ListNode));
    struct ListNode* p=l1;
    struct ListNode* q=l2;
    struct ListNode* r=result;
    int sum, residue=0, dec;

    while (p || q || residue){
        sum=residue;
        if (p){
            sum+=p->val;
            p=p->next;
        }
        if (q){
            sum+=q->val;
            q=q->next;
        }
        residue=sum/10;
        r->val=sum%10;
        if (p || q || residue){
            r->next=(struct ListNode*)malloc(sizeof(struct ListNode));
            r=r->next;
        }
        r->next=NULL;
    }
    return result;
}